tests/integration/conftest.py

```diff
@@ -82,43 +128,62 @@ def port_is_listening(hostname: str, port: int, ipv6: bool) -> bool:
 
 
 def wait_for_kafka(port: int, *, hostname: str = "127.0.0.1", wait_time: float = 20.0) -> None:
-    start_time = time.monotonic()
     bootstrap_server = f"{hostname}:{port}"
-    while True:
-        if time.monotonic() - start_time > wait_time:
-            raise Timeout(f"Could not contact kafka cluster on host {hostname}, port {port}")
+    expiration = Expiration.from_timeout(
+        msg=f"Could not contact kafka cluster on host `{bootstrap_server}`",
+        timeout=wait_time,
+    )
+
+    list_topics_successful = False
+    while not list_topics_successful:
+        expiration.raise_if_expired()
         try:
-            client = KafkaAdminClient(bootstrap_servers=bootstrap_server)
-            _ = client.list_topics()
-            return
-        # pylint: disable=broad-except
-        except Exception as e:
+            KafkaAdminClient(bootstrap_servers=bootstrap_server).list_topics()
+        except Exception as e:  # pylint: disable=broad-except
             print(f"Error checking kafka cluster: {e}")
             time.sleep(2.0)
+        else:
+            list_topics_successful = True
 
 
 def wait_for_port(port: int, *, hostname: str = "127.0.0.1", wait_time: float = 20.0, ipv6: bool = False) -> None:
     start_time = time.monotonic()
-    while True:
-        if port_is_listening(hostname, port, ipv6):
-            break
-        if time.monotonic() - start_time > wait_time:
-            raise Timeout("Timeout waiting for port {} on host {}".format(port, hostname))
+    expiration = Expiration(
+        msg=f"Timeout waiting for `{hostname}:{port}`",
+        deadline=start_time + wait_time,
+    )
+
+    while not port_is_listening(hostname, port, ipv6):
+        expiration.raise_if_expired()
         time.sleep(2.0)
-    print("Port {} on host {} started listening in {} seconds".format(port, hostname, time.monotonic() - start_time))
 
+    elapsed = time.monotonic() - start_time
+    print(f"Server `{hostname}:{port}` listening after {elapsed} seconds")
+
+
+def get_random_port(*, port_range: PortRangeInclusive, blacklist: List[int]) -> int:
+    """ Find a random port in the range `PortRangeInclusive`.
 
-def get_random_port(*, start: int = 3000, stop: int = 30000, blacklist: Optional[List[int]] = None) -> int:
-    while True:
-        value = random.randrange(start, stop)
-        if not blacklist or value not in blacklist:
-            return value
+    Note:
+        This function is *not* aware of the ports currently open in the system,
+        the blacklist only prevents two services of the same type to randomly
+        get the same ports for *a single test run*.
+
+        Because of that, the port range should be chosen such that there is no
+        system service in the range. Also note that running two sessions of the
+        tests with the same range is not supported and will lead to flakiness.
+    """
+    value = random.randint(port_range.start, port_range.end)
+    while value in blacklist:
```

Comment:
[Small note] It looks for me that in a scenarios I saw later, blacklist is not so necessary: As far as I know two adjacent calls or random.randint() return the same value with a very low probability (if range is wide, of course).

This approach is totally fine, but I would rather write a global generator in `BeforeAll` and  withdraw next available port each time. Totally up to you.

Reply:
> As far as I know two adjacent calls or random.randint() return the same value with a very low probability (if range is wide, of course).

Using a deterministic series with global states breaks with parallelization, and I'm trying to go towards parallelization to speed up the tests.

This code is far from perfect, there is actually a non-negligible chance of getting a conflict, specially because of parallelization and some repeated runs. Ideally the start port number should be provided externally to the test runner, then we would draw the ports in the fashion you described, and even then try to check if the port is bound before using it.

I just didn't want to introduce these additional changes in this PR that was getting big already.